lib/dry/result_matcher/evaluator.rb

```diff
@@ -0,0 +1,40 @@
+module Dry
+  module ResultMatcher
+    class Evaluator
+      attr_reader :result
+      attr_reader :output
+
+      def initialize(result, cases)
+        @result = result
+        @output = nil
+        @matched = false
+
+        cases.each do |name, kase|
+          define_singleton_method(name) do |pattern = nil, &block|
+            handle_case(kase, pattern, &block)
+          end
+        end
```

Comment:
Gotcha, I'm almost 100% sure method_missing will have better performance => [see what happens when you invalidate method cache](http://dev.mensfeld.pl/2015/04/ruby-global-method-cache-invalidation-impact-on-a-single-and-multithreaded-applications/)